cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceRuntimeService.java

```diff
@@ -471,9 +474,18 @@ public Void call() throws Exception {
         try {
           mapReduce.beforeSubmit(context);
 
-          // set input/output datasets info
-          setInputDatasetIfNeeded(job);
-          setOutputDatasetsIfNeeded(job);
+          TypeToken<?> mapperTypeToken = checkMapperOutputTypes(job);
+
+          // set input/outputs info
+          setInputsIfNeeded(job);
+          setOutputsIfNeeded(job);
+
+          if (mapperTypeToken == null) {
+            // means there was no inputs configured through CDAP APIs.
```

Comment:
`checkMapperOutputTypes` returning null indicates that user didn't configure inputs through CDAP APIs, so here we set mapperTypeToken based upon the value from the job (likely the identity mapper).